xla/service/elemental_ir_emitter.cc

```diff
@@ -793,6 +793,123 @@ llvm::Value* EmitF8e4m3b11fnuzToF16(llvm::Value* f8_value,
   return f16_value;
 }
 
+absl::StatusOr<llvm::Value*> EmitF16ToF4e2m1fn(llvm::Value* f16_value,
+                                               llvm::IRBuilder<>* b) {
```

Comment:
Like in EmitF16ToF8e, can you add comments that show the pseduocode of what the LLVM op does? E.g. before defining `llvm::Value* as_int8`, you could have the comment

```
as_int8 = as_int16 >> 9
```

Also like in EmitF16ToF8e, having a few explanation comments and constants would help. E.g. instead of the constant 28 below when defining f4_normal, you could have the constant `mantissa_bits_difference << f4_mantissa_bits`.

It might be worth reusing EmitF16ToF8e for F4 as well. You'd have to handle the fact there is no Inf though, as well as modify handle_halfway_points_F16ToF8, so this might not be worth it. Even if not reusing EmitF16ToF8e, it might be worth copying the general structure for consistency, as most of the code there applies to F4 as well.

And similar comments apply to EmitF4e2m1fnToF16, where you can potentially reuse EmitToF16F8e.